url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/block/BlockTriangularSolver.java

```diff
@@ -0,0 +1,516 @@
+/*
+ * Copyright (c) 2009-2013, Peter Abeles. All Rights Reserved.
+ *
+ * This file is part of Efficient Java Matrix Library (EJML).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ejml.alg.block;
+
+import org.ejml.data.D1Submatrix64F;
+
+import static org.ejml.alg.block.BlockInnerMultiplication.blockMultMinus;
+
+
+/**
+ * <p>
+ * Contains triangular solvers for {@link org.ejml.data.BlockMatrix64F} block aligned sub-matrices.
+ * </p>
+ *
+ * <p>
+ * For a more detailed description of a similar algorithm see:
+ * Page 30 in "Fundamentals of Matrix Computations" 2nd Ed. by David S. Watkins
+ * or any description of a block triangular solver in any other computational linear algebra book.
+ * </p>
+ *
+ * @author Peter Abeles
+ */
+public class BlockTriangularSolver {
+
+    /**
+     * Inverts an upper or lower triangular block submatrix.
+     *
+     * @param blockLength
+     * @param upper Is it upper or lower triangular.
+     * @param T Triangular matrix that is to be inverted.  Must be block aligned.  Not Modified.
+     * @param T_inv Where the inverse is stored.  This can be the same as T.  Modified.
+     * @param temp Work space variable that is size blockLength*blockLength.  
+     */
+    public static void invert( final int blockLength ,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.